Modules/_asynciomodule.c

```diff
@@ -74,8 +78,19 @@ typedef struct {
 #define Future_CheckExact(state, obj) Py_IS_TYPE(obj, state->FutureType)
 #define Task_CheckExact(state, obj) Py_IS_TYPE(obj, state->TaskType)
 
-#define Future_Check(state, obj) PyObject_TypeCheck(obj, state->FutureType)
-#define Task_Check(state, obj) PyObject_TypeCheck(obj, state->TaskType)
+#define Future_Check(state, obj)                        \
+    (Future_CheckExact(state, obj)                      \
+     || PyObject_TypeCheck(obj, state->FutureType))
+
+#define Task_Check(state, obj)                          \
+    (Task_CheckExact(state, obj)                        \
+     || PyObject_TypeCheck(obj, state->TaskType))
+
+#define TaskOrFuture_Check(state, obj)                  \
+    (Task_CheckExact(state, obj)                        \
+     || Future_CheckExact(state, obj)                   \
+     || PyObject_TypeCheck(obj, state->FutureType)      \
+     || PyObject_TypeCheck(obj, state->TaskType))
```

Comment:
```suggestion
#define TaskOrFuture_Check(state, obj)                  \
    (Task_Check(state, obj)                             \
     || Future_Check(state, obj)                   
```

Reply:
Kumar, I'm not applying this suggestion, because the current code is optimized for a fast path. With your suggestion the slow `PyObject_TypeCheck` will still be called for native Futures and I want to avoid that by inlining a faster check before that.